mesonbuild/mtest.py

```diff
@@ -1849,36 +1863,52 @@ def test_suitable(self, test: TestSerialisation) -> bool:
 
         return True
 
-    def tests_from_args(self, tests: T.List[TestSerialisation]) -> T.Generator[TestSerialisation, None, None]:
+    @staticmethod
+    def make_selection_checker(patterns: T.List[str], should_match_if_no_patterns: bool = False) -> T.Callable[[TestSerialisation], bool]:
         '''
-        Allow specifying test names like "meson test foo1 foo2", where test('foo1', ...)
+        Allow specifying test names like "foo1", which will match test('foo1', ...)
 
         Also support specifying the subproject to run tests from like
-        "meson test subproj:" (all tests inside subproj) or "meson test subproj:foo1"
-        to run foo1 inside subproj. Coincidentally also "meson test :foo1" to
+        "subproj:" (all tests inside subproj) or "subproj:foo1"
+        to run foo1 inside subproj. Coincidentally also ":foo1" to
         run all tests with that name across all subprojects, which is
-        identical to "meson test foo1"
+        identical to "foo1"
         '''
-        for arg in self.options.args:
-            if ':' in arg:
-                subproj, name = arg.split(':', maxsplit=1)
+        if not patterns:
+            return lambda _: should_match_if_no_patterns
+
+        unqualified = set()  # type: T.Set[str]
+        qualified = {}  # type: T.Dict[str, T.Union[T.Set[str], UniversalSet[str]]]
+
+        for pattern in patterns:
+            if ':' in pattern:
+                subproj, name = pattern.split(':', maxsplit=1)
             else:
-                subproj, name = '', arg
-            for t in tests:
-                if subproj and t.project_name != subproj:
-                    continue
-                if name and t.name != name:
-                    continue
-                yield t
+                subproj, name = '', pattern
+
+            if subproj == '':
+                unqualified.add(name)
+            elif name == '':
+                qualified[subproj] = UniversalSet()
+            else:
+                selected_subproj_tests = qualified.get(subproj, set())
+                selected_subproj_tests.add(name)
+
+        def is_selected(t: TestSerialisation) -> bool:
+            selected_subproj_tests = qualified.get(t.project_name, set())
+
+            return (t.name in selected_subproj_tests or t.name in unqualified)
```

Comment:
With `defaultdict` this would become `return t.name in qualified[t.project_name] or t.name in unqualified`. Since it would be a one line, maybe this could become a lambda again?